addons/test_mail/tests/test_mail_followers_unsubscribe.py

```diff
@@ -46,3 +53,148 @@ def test_inbox_notification_follower(self):
+        cls.test_record_unfollow = cls.env['mail.test.simple.unfollow'].create({'name': 'unfollow'})
+        cls.test_record_unfollow_admin = cls.test_record_unfollow.with_user(cls.user_admin)
+        cls.user_employee.write({'notification_type': 'email'})
+        cls.partner_admin = cls.user_admin.partner_id
+        cls.partner_without_user = cls.env['res.partner'].create({
+            'name': 'Dave',
+            'email': 'dave@odoo.com',
+        })
+
+    def test_initial_data(self):
+        """Test some initial value."""
+        self.assertTrue(self.user_employee._is_internal())
+        record_employee = self.test_record.with_user(self.user_employee)
+        record_employee.check_access_rights('read')
+        record_employee.check_access_rule('read')
+        mail_template_arch = self.env.ref('mail.mail_notification_layout').arch
+        self.assertIn('/mail/unfollow', mail_template_arch)
+        self.assertNotIn('/mail/unfollow', re.sub(_UNFOLLOW_REGEX, '', mail_template_arch))
+
+    def _extract_unfollow_url(self, mail_body):
+        urls = {link_url for _, link_url, _, _ in re.findall(tools.HTML_TAG_URL_REGEX, mail_body)
+                if '/mail/unfollow' in link_url}
+        n_url = len(urls)
+        self.assertLessEqual(n_url, 1)
+        return next(iter(urls), None)
+
+    def _message_post(self, record):
+        with self.mock_mail_gateway():
+            record.message_post(body='test message', subtype_id=self.env.ref('mail.mt_comment').id)
+            self.assertEqual(len(self._mails), 1)
+            return self._mails[0]['body']
+
+    def _notify_thread_by_email(self, record, partner):
+        with self.mock_mail_gateway():
+            recipients_data = self._generate_notify_recipients(partner)
+            test_message = self.env['mail.message'].browse(self.test_message.ids).copy()
+            record._notify_thread_by_email(test_message, recipients_data)
+            self.assertEqual(len(self._mails), 1)
+            return self._mails[0]['body']
+
+    def _url_tampering(self, url, **kwargs):
+        parsed_url = urlparse(url)
+        return parsed_url._replace(query=urlencode(dict(parse_qsl(parsed_url.query), **kwargs))).geturl()
+
+    def _test_tampered_unfollow_url(self, record, unfollow_url, partner):
+        for param, value in (('token', '0000000000000000000000000000000000000000'),
+                             ('model', 'mail.test.gateway'),
+                             ('res_id', record.copy().id)):
+            with self.subTest(f'Tampered {param}'):
+                tampered_unfollow_url = self._url_tampering(unfollow_url, **{param: value})
+                response = self.url_open(tampered_unfollow_url)
+                self.assertEqual(response.status_code, 403)
+                self.assertIn(partner, record.message_partner_ids)
+
+        with self.subTest('Tampered partner id'):
+            record._message_subscribe(partner_ids=self.partner_admin.ids)
```

Comment:
There are a lot of nested tests ... a test doing business code, calling a tool method, doing subtests, calling other business methods, then other asserts, ... have to say testes are not clear to me ^^ can't we be a bit more straightforward even if it is a bit more verbose ?

Reply:
Actually, I have rewritten the test and done correction following your advices above. I have also added comments to the test to ease understanding. I think it is now easier to understand (also because there are less strange think, thanks to you). 
I think that if I unroll the sub test  _test_unfollow_url, _test_tampered_unfollow_url which test respectively if the link works and that the tampered one doesn’t work it would be very long and maybe would complexify the test. Would you mind to take a look at the new version and see if it now more easy to understand. If not, I will try to unroll some method.